### PR TITLE
Export security-analysis results

### PR DESCRIPTION
**Pull request by klesaulnier**

### CODE REVIEW

**Review by Meklo — APPROVED** [review]

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 1]
```diff
@@ -596,11 +607,131 @@ public void testSecurityAnalysisParameters() throws Exception {
...
+         * RUN SECURITY ANALYSIS END
+         */
+
+        // get NOT_FOUND security analysis result zipped csv
+        mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/security-analysis/result/csv?resultType={resultType}", studyUuid, nodeUuid, SecurityAnalysisResultType.N).content(CSV_TRANSLATION_DTO_STRING)).andExpectAll(
+            status().isNotFound());
+
+        assertTrue(TestUtils.getRequestsWithBodyDone(1, server).stream().anyMatch(r ->
+            r.getPath().matches(String.format("/v1/results/%s/n-result/csv", SECURITY_ANALYSIS_OTHER_NODE_RESULT_UUID))
+                && r.getBody().equals(CSV_TRANSLATION_DTO_STRING)
```
Comment (Tristan-WorkGH) [inline src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:720]:
```suggestion
                && CSV_TRANSLATION_DTO_STRING.equals(r.getBody())
```

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 2]
```diff
@@ -596,11 +607,131 @@ public void testSecurityAnalysisParameters() throws Exception {
...
+
+        // get NMK_CONSTRAINTS security analysis result zipped csv
+        mvcResult = mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/security-analysis/result/csv?resultType={resultType}", studyUuid, nodeUuid, SecurityAnalysisResultType.NMK_LIMIT_VIOLATIONS).content(CSV_TRANSLATION_DTO_STRING)).andExpectAll(
+            status().isOk()).andReturn();
+        byteArrayResult = mvcResult.getResponse().getContentAsByteArray();
+        assertTrue(Arrays.equals(SECURITY_ANALYSIS_NMK_CONSTRAINTS_RESULT_CSV_ZIPPED, byteArrayResult));
+
+        assertTrue(TestUtils.getRequestsWithBodyDone(1, server).stream().anyMatch(r ->
+            r.getPath().matches(String.format("/v1/results/%s/nmk-constraints-result/csv", SECURITY_ANALYSIS_RESULT_UUID))
+                && r.getBody().equals(CSV_TRANSLATION_DTO_STRING)
```
Comment (Tristan-WorkGH) [inline src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:675]:
```suggestion
                && CSV_TRANSLATION_DTO_STRING.equals(r.getBody())
```

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 3]
```diff
@@ -596,11 +607,131 @@ public void testSecurityAnalysisParameters() throws Exception {
...
+
+        // get NMK_CONTINGENCIES security analysis result zipped csv
+        mvcResult = mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/security-analysis/result/csv?resultType={resultType}", studyUuid, nodeUuid, SecurityAnalysisResultType.NMK_CONTINGENCIES).content(CSV_TRANSLATION_DTO_STRING)).andExpectAll(
+            status().isOk()).andReturn();
+        byteArrayResult = mvcResult.getResponse().getContentAsByteArray();
+        assertTrue(Arrays.equals(SECURITY_ANALYSIS_NMK_CONTINGENCIES_RESULT_CSV_ZIPPED, byteArrayResult));
+
+        assertTrue(TestUtils.getRequestsWithBodyDone(1, server).stream().anyMatch(r ->
+            r.getPath().matches(String.format("/v1/results/%s/nmk-contingencies-result/csv", SECURITY_ANALYSIS_RESULT_UUID))
+                && r.getBody().equals(CSV_TRANSLATION_DTO_STRING)
```
Comment (Tristan-WorkGH) [inline src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:664]:
```suggestion
                && CSV_TRANSLATION_DTO_STRING.equals(r.getBody())
```

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 4]
```diff
@@ -596,11 +607,131 @@ public void testSecurityAnalysisParameters() throws Exception {
...
+
+        // get N security analysis result zipped csv
+        MvcResult mvcResult = mockMvc.perform(post("/v1/studies/{studyUuid}/nodes/{nodeUuid}/security-analysis/result/csv?resultType={resultType}", studyUuid, nodeUuid, SecurityAnalysisResultType.N).content(CSV_TRANSLATION_DTO_STRING)).andExpectAll(
+            status().isOk()).andReturn();
+        byte[] byteArrayResult = mvcResult.getResponse().getContentAsByteArray();
+        assertTrue(Arrays.equals(SECURITY_ANALYSIS_N_RESULT_CSV_ZIPPED, byteArrayResult));
+
+        assertTrue(TestUtils.getRequestsWithBodyDone(1, server).stream().anyMatch(r ->
+            r.getPath().matches(String.format("/v1/results/%s/n-result/csv", SECURITY_ANALYSIS_RESULT_UUID))
+                && r.getBody().equals(CSV_TRANSLATION_DTO_STRING)
```
Comment (Tristan-WorkGH) [inline src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:653]:
```suggestion
                && CSV_TRANSLATION_DTO_STRING.equals(r.getBody())
```

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 5]
```diff
@@ -792,6 +792,18 @@ public ResponseEntity<String> getSecurityAnalysisResult(@Parameter(description =
...
 
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/security-analysis/result/csv")
+    @Operation(summary = "Get a security analysis result on study - CSV export")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The security analysis result csv export"),
+        @ApiResponse(responseCode = "204", description = "No security analysis has been done yet"),
+        @ApiResponse(responseCode = "404", description = "The security analysis has not been found")})
+    public byte[] getSecurityAnalysisResult(@Parameter(description = "study UUID") @PathVariable("studyUuid") UUID studyUuid,
+                                                                           @Parameter(description = "nodeUuid") @PathVariable("nodeUuid") UUID nodeUuid,
+                                                                           @Parameter(description = "result type") @RequestParam(name = "resultType") SecurityAnalysisResultType resultType,
+                                                                           @Parameter(description = "Csv translation") @RequestBody String csvTranslations) {
```
[the remark this answers is not in the document]
Comment (klesaulnier) [reply src/main/java/org/gridsuite/study/server/StudyController.java:803]: fixed

File: src/main/java/org/gridsuite/study/server/service/SecurityAnalysisService.java [hunk 6]
```diff
@@ -81,7 +81,7 @@ public String getSecurityAnalysisResult(UUID nodeUuid, SecurityAnalysisResultTyp
             return null;
         }
 
-        UriComponentsBuilder pathBuilder = UriComponentsBuilder.fromPath(DELIMITER + SECURITY_ANALYSIS_API_VERSION + "/results/{resultUuid}/" + getPathFromResultType(resultType))
+        UriComponentsBuilder pathBuilder = UriComponentsBuilder.fromPath(DELIMITER + SECURITY_ANALYSIS_API_VERSION + "/results/{resultUuid}/" + getPagedPathFromResultType(resultType))
             .queryParam("page", pageable.getPageNumber())
             .queryParam("size", pageable.getPageSize());
```
[the remark this answers is not in the document]
Comment (Tristan-WorkGH) [reply src/main/java/org/gridsuite/study/server/service/SecurityAnalysisService.java:86]: ah no ... nothing doing it for us

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 7]
```diff
@@ -205,6 +205,17 @@ public MockResponse dispatch(RecordedRequest request) {
                 } else if (path.matches("/v1/results/" + SECURITY_ANALYSIS_RESULT_UUID + "/nmk-constraints-result/paged\\?page=.*size=.*filters=.*sort=.*")) {
                     return new MockResponse().setResponseCode(200).setBody(SECURITY_ANALYSIS_NMK_CONSTRAINTS_RESULT_JSON)
                         .addHeader("Content-Type", "application/json; charset=utf-8");
+                } else if (path.matches("/v1/results/" + SECURITY_ANALYSIS_RESULT_UUID + "/n-result/csv")) {
+                    return new MockResponse().setResponseCode(200).setBody(getBinaryAsBuffer(SECURITY_ANALYSIS_N_RESULT_CSV_ZIPPED))
+                        .addHeader("Content-Type", "application/json; charset=utf-8");
```
[the remark this answers is not in the document]
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:210]: fixed in the whole file

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 8]
```diff
@@ -205,6 +205,17 @@ public MockResponse dispatch(RecordedRequest request) {
...
+                } else if (path.matches("/v1/results/" + SECURITY_ANALYSIS_OTHER_NODE_RESULT_UUID + "/n-result/csv")) {
+                    return new MockResponse().setResponseCode(404);
+                } else if (path.matches("/v1/results/" + SECURITY_ANALYSIS_RESULT_UUID + "/nmk-contingencies-result/csv")) {
+                    return new MockResponse().setResponseCode(200).setBody(getBinaryAsBuffer(SECURITY_ANALYSIS_NMK_CONTINGENCIES_RESULT_CSV_ZIPPED))
+                        .addHeader("Content-Type", "application/json; charset=utf-8");
```
[the remark this answers is not in the document]
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:215]: same

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java [hunk 9]
```diff
@@ -205,6 +205,17 @@ public MockResponse dispatch(RecordedRequest request) {
...
+                } else if (path.matches("/v1/results/" + SECURITY_ANALYSIS_RESULT_UUID + "/nmk-constraints-result/csv")) {
+                    return new MockResponse().setResponseCode(200).setBody(getBinaryAsBuffer(SECURITY_ANALYSIS_NMK_CONSTRAINTS_RESULT_CSV_ZIPPED))
+                        .addHeader("Content-Type", "application/json; charset=utf-8");
```
[the remark this answers is not in the document]
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:218]: same

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java (hunk 1 above)
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:720]: fixed

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java (hunk 2 above)
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:675]: fixed

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java (hunk 3 above)
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:664]: fixed

File: src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java (hunk 4 above)
Comment (klesaulnier) [reply src/test/java/org/gridsuite/study/server/SecurityAnalysisTest.java:653]: fixed

**Review by Tristan-WorkGH — APPROVED** [review]

**Pull request merged**